BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,437 @@
+
+    assert(output);
+
+    auto buffer = static_cast<Baikal::ClwOutput*>(output)->data();
+    output_data.resize(buffer.GetElementCount());
+    image_data.resize(buffer.GetElementCount());
+
+    output->GetData(output_data.data());
+
+    TypeDesc fmt;
+    switch (bpp)
+    {
+        case 8:
+            fmt = TypeDesc::UINT8;
+            break;
+        case 16:
+            fmt = TypeDesc::UINT16;
+            break;
+        case 32:
+            fmt = TypeDesc::FLOAT;
+            break;
+        default:
+            throw std::runtime_error("Render::SaveOutput(...):"
+                                     "Unhandled bpp of image.");
+    }
+
+    for (auto y = 0u; y < height; ++y)
+    {
+        for (auto x = 0u; x < width; ++x)
+        {
+            float3 val = output_data[(height - 1 - y) * width + x];
+            val *= (1.f / val.w);
+            image_data[y * width + x].x = std::pow(val.x, 1.f / 2.2f);
+            image_data[y * width + x].y = std::pow(val.y, 1.f / 2.2f);
+            image_data[y * width + x].z = std::pow(val.z, 1.f / 2.2f);
+        }
+    }
+
+    std::stringstream ss;
+
+    auto time = std::chrono::high_resolution_clock::now();
+
+    ss << path << "/" << file_name << "-" 
+       << std::to_string(time.time_since_epoch().count())
+       << "." << extension;
+
+    std::unique_ptr<ImageOutput> out(ImageOutput::create(ss.str()));
+
+    if (!out)
+    {
+        throw std::runtime_error("Render::SaveOutput(...):"
+                                 "Can't create image file on disk");
+    }
+
+    ImageSpec spec(width, height, 3, TypeDesc::FLOAT);
+    out->open(ss.str(), spec);
+    out->write_image(TypeDesc::FLOAT, image_data.data(), sizeof(float3));
+    out->close();
+}
+
+void Render::GenerateDataset(const std::string &path)
+{
+    using namespace RadeonRays;
+
+    for (const auto &cam_state: m_camera_states)
+    {
+        if (!m_camera)
+        {
+            m_camera = Baikal::PerspectiveCamera::Create(cam_state.at, cam_state.pos, cam_state.up);
+            m_scene->SetCamera(m_camera);
+            m_camera->SetSensorSize(RadeonRays::float2(0.036f, 0.036f));
+            m_camera->SetDepthRange(RadeonRays::float2(0.0f, 100000.f));
+
+            m_camera->SetSensorSize(RadeonRays::float2(0.036f, 0.036f));
+            m_camera->SetDepthRange(RadeonRays::float2(0.0f, 100000.f));
+            m_camera->SetFocalLength(0.035f);
+            m_camera->SetFocusDistance(1.f);
+            m_camera->SetAperture(0.f);
+
+            m_scene->SetCamera(m_camera);
+        }
+
+        UpdateCameraSettings(cam_state);
+
+        for (const auto& output: m_outputs)
+        {
+            output->Clear(RadeonRays::float3(.0f, .0f, .0f, .0f));
+        }
+
+        m_controller->CompileScene(m_scene);
+        auto& scene = m_controller->GetCachedScene(m_scene);
+
+        for (auto i = 0u; i < kNumIterations; i++)
+        {
+            m_renderer->Render(scene);
```

Comment:
We should save outputs not only for target value, but also for intermediate samples count: 2, 4, 8, 16, etc. It should be configurable

Reply:
done